Add markdown documentation for fractribution dbt models

zloff opened the pull request:
- I wasn't sure what stage of the tutorial this should be in - I have added it as a 'next step' for now.
- How will fractribution be integrated? Is it going to be a separate dbt package, or will it be integrated into dbt-snowplow? I have added placeholder paths showing the location of the macros but will need to update these depending on where the files sit in the directory structure

miike commented:
On steps:
- We'll replace all the current steps that are there as these are only really relevant for specifically the web accelerator.
- Our steps will be reasonably similar however:
1. Sample data (either upload or data share, fine to leave as upload for the moment)
2. Modelling (we assume the snowplow-dbt-web package has already been run. We also assume that the fractribution model is a separate dbt package entirely from snowplow-dbt-web.
3. Modelling will likely need to go into some level of depth around the customisations required of fractribution as well as how to run it.
4. For visualisation we'll just output an example table - this can be as HTML / markdown rather than an image.
5. For tracking - let's keep this section but we really only need two key events for fractribution - page_views - so that we can determine where the user came from at the start of the session (or throughout the session) and attribute it to a marketing channel and conversions - which is just an event (or events) where there's a transaction value attached. With those two events we can then tie the transaction to the marketing channel(s) that the user has visited and calculate the relative contribution of each channel (for all conversions) across all conversion channels. This is then joined to the channel spend to calculate the return on investment i.e., for a given amount of money spent on a channel (effectively advertising) what is the returned value in terms of transactions?
6. For enrich we only really need one enrichment setup which is pretty basic which is the campaign attribution enrichment https://docs.snowplow.io/docs/enriching-your-data/available-enrichments/campaign-attribution-enrichment/ which takes the UTMs in a given URL and splits them out into seperate columns in Snowplow that can then be used for analysis.

miike commented on content/modeling/modeling_1.md:
````diff
@@ -1,16 +1,16 @@
 
 ```yml
 packages:
-  - package: snowplow/snowplow_web
````
I think maybe we keep this as we're assuming that the user has either run the snowplow_web dbt model independently. There may be some users that haven't in which case keeping this in packages may make sense.

miike commented on content/modeling/modeling_3.md:
```diff
@@ -1,81 +1,22 @@
 +++
 
-Data should now be loaded into your warehouse. In this section we will take a closer look at the output to mitigate data issues and get familiar with the derived tables.
+Data from the fractribution dbt package should now be loaded into your warehouse. Now we need to create the report table. There are two ways to to this - with Python, or by pulling a Docker container image. 
```
Python => `Python locally`

miike commented on content/modeling/modeling_3.md:
```diff
@@ -1,81 +1,22 @@
-1. [your_custom_schema]_***scratch***: drop and recompute models that aid the incremental run
-2. [your_custom_schema]_***derived***: main output models you can use in your downstream models and reporting
-3. [your_custom_schema]_***snowplow_manifest***: tables that help the integrity and core incremental logic of the model
+{{< tabs groupId="modeling" >}}
```
Is there enough info on how to run this locally at the moment based on the code i pushed?

zloff replied:
It's a bit of a struggle, but I'll do my best to write up what I can

zloff replied:
I see you're using snowflake.snowpark - there's a note in the docs about errors running on M1 and specific instructions for the virtual environment: https://docs.snowflake.com/en/developer-guide/snowpark/python/setup.html
I'll include this in the setup instructions, i guess

zloff commented on content/modeling/modeling_2.md:
```diff
@@ -1,51 +1,115 @@
+- Unique: all events in a path are treated as unique (no reduction of complexity). Best for smaller datasets (small lookback window) without a lot of retargeting
+- First: keep only the first occurrence of any event: A -> B -> A becomes A -> B. Best for brand awareness marketing.
+- Frequency: keep a count of the events' frequency: A -> A -> B becomes A(2) -> B. Best when there is a lot of retargeting
+- Recency: bucket events depending on when they occurred before the conversion and collapse those in the same time bucket. Best for larger lookback windows
```
Ah yep, I'll add something

zloff commented:
-For the modelling_3, I gave it a shot, but not too sure exactly what instructions to give for docker. Would the docker image come with everything they need, so that they just do a docker run with the snowflake configs? I also wasn't sure if I should instead pass in an environment vars file instead and get them to create/update that?

-I gave the tracking part my best shot, but had trouble figuring out how to instrument the transaction event, especially in the React and Angular tabs in tracking_2 -(from line146 onwards) i can't find any mentions in the snowplow docs about enableEcommerceTracking() in a similar style to what was previously there with link clicks etc, so I just guessed. Also, the Angular tab (end of the file) has an Add tracking to App step, but React above it doesn't? It's hard to write a tutorial for something you don't know much about haha

- I left tracking_3 and tracking_4 alone, not sure if we want to keep tracking_3?

miike commented on content/modeling/modeling_3.md:
```diff
@@ -1,81 +1,103 @@
+- last_touch: Assigns 100% attribution to the last channel in each path.
+- position_based: The first and last channels get 40% of the credit each, with the remaining
+    channels getting the leftover 20% distributed evenly.
+- linear: Assigns attribution evenly between all channels on the path.
```
Let's add a visual image here to demonstrate what is going on as well. We can re-use this image (and link back to the source) here: https://www.thinkwithgoogle.com/intl/en-ca/marketing-strategies/data-and-measurement/measurement-and-attribution-primer/

zloff commented:
Hey Mike, 
I've added a folder with the Dockerfile and fractribution files - i figure we will remove it eventually but this is just so that you can see what i've written up so far. I wasn't sure about the attribution_model flag - I wasn't sure how to run a different CMD if attribution_model is set in the .env file vs when it isn't. To get around it, i've moved the default model logic to the Dockerfile (`ENV attribution_model=${attribution_model:-shapley}`) which is maybe not good practice?

_I'd be tempted to just stick with Javascript for the moment and leave React / Angular as is._
I wasn't sure what you meant for me to do here, do i delete those tabs?

miike commented on fractribution_docker/main_snowplow_snowflake.py:
```diff
@@ -0,0 +1,362 @@
+FLAGS = flags.FLAGS
+
+connection_parameters = {
+    # "account": os.environ["snowflake_account"],
```
Does this need to be uncommented now?

zloff replied:
I figured that this code was your domain - I haven't done anything with it. Was I supposed to? 
Reading through, I saw that there were a bunch of comments and questions, I figured it wasn't finished yet

miike replied:
Maybe just make some minor changes? I want to ensure we can also get this running in / on your machine as well.

miike commented on fractribution_docker/main_snowplow_snowflake.py:
```diff
@@ -0,0 +1,362 @@
+
+  # assumes that the dataset already exists
+  params['channel_counts_table'] = 'channel_counts'
+  params['attribution_model'] = 'shapley'
```
Do we read in from the command line? This looks like it uses shapley irrespective of what is passed in as the parameter?

zloff replied:
Yeah i find the input_params with _get_template_params and _get_param_or_die functions a bit confusing. I don't think it's set up right - the script runs fine when i run `python fractribution.py --attribution_model dhrtrhf`

zloff replied:
Okay have fixed this now to use the provided parameter, or default if none provided

zloff commented:
Next step @miike - did you have any ideas for how you want to handle the `conversion_window_start_date` and `conversion_window_end_date`? They're currently fixed in the script. We could alter a couple of the DBT models to include these as two extra columns, and then grab them when making the report_table?

miike commented:
Snowplow now have a template we can use that they sent me last night - https://docs.snowplow.io/accelerators/template/

miike commented on content/modeling/modeling_3.md:
````diff
@@ -1,81 +1,120 @@
-**Find out details about the highest engaged user using `derived.snowplow_users:`**
-```sql
+```
+python main_snowplow_snowflake.py --conversion_window_start_date '2022-07-03' --conversion_window_end_date '2022-08-01' --attribution_model linear
````
Let's default to shapley here in the command rather than linear.

zloff replied:
Ah I added something other than shapley because this was to demonstrate how to use something other than the default (shapley) by adding a flag. But yeah I see how people could just not read that and copy paste it. I'll update

miike commented on content/modeling/modeling_3.md:
```diff
@@ -1,81 +1,120 @@
+conversion_window_start_date='2022-07-03'
+conversion_window_end_date='2022-08-01'
+# Add a variable if you wish to change the attribution model, e.g.:
+attribution_model=linear
```
shapley

miike commented on content/modeling/modeling_3.md:
````diff
@@ -1,81 +1,120 @@
 
-SELECT * FROM ENGAGEMENT
 ```
+docker run --env-file /path/to/env/file/configs.env -it snowplow/fractribution 
````
Let's add `--rm` flag to automatically remove the container after running (this way if a user runs it multiple times it doesn't create multiple containers).

miike commented on fractribution_docker/main_snowplow_snowflake.py:
```diff
@@ -0,0 +1,355 @@
+  return (len(column_name) <= 300 and
+          VALID_CHANNEL_NAME_PATTERN.match(column_name) is not None)
+
+# ZL not using this func anymore
```
Let's remove it then.

miike commented on fractribution_docker/main_snowplow_snowflake.py:
```diff
@@ -0,0 +1,355 @@
+      raise ValueError('Channel is not a legal Snowflake column name: ', channel)
+  return channels
+
+# ZL there's a lot that we can delete in this func
```
the original frac code has this but I don't think we really need it (to the same extent of templating).

If possible let's move towards using the `argparse` library to input and parse parameters instead. This is a more complex example (we don't need this many args) but can be used as a reference - https://github.com/poplindata/snowplow-event-generation/blob/master/generate.py#L837

miike commented on fractribution_docker/main_snowplow_snowflake.py:
```diff
@@ -0,0 +1,355 @@
+  print(report)
+  
+  session.close()
+  # TODO: need to join (channel_attribution) to spend data
```
Can probably delete the comments below as I think we already create this report table above.

miike commented on fractribution_docker/main_snowplow_snowflake.py:
```diff
@@ -0,0 +1,355 @@
+  return 0
+
+
+def main(event, unused_context=None) -> int:
```
We can delete this as we're assuming the user doesn't want to deploy this as a cloud function.
